dacite/core.py

```diff
@@ -91,15 +92,25 @@ def _build_value_for_union(union: Type, data: Any, config: Config) -> Any:
     types = extract_generic(union)
     if is_optional(union) and len(types) == 2:
         return _build_value(type_=types[0], data=data, config=config)
+    valid_resolutions = {}
     for inner_type in types:
         try:
             value = _build_value(type_=inner_type, data=data, config=config)
             if is_instance(value, inner_type):
-                return value
+                if not config.no_ambiguous_resolution:
+                    return value
+                else:
+                    valid_resolutions[inner_type] = value
         except DaciteError:
             pass
-    if not config.check_types:
-        return data
+    if config.no_ambiguous_resolution:
+        if len(valid_resolutions) > 1:
+            raise AmbiguousResolutionError(valid_resolutions)
+        else:
+            return valid_resolutions.popitem()[1]
+    else:
```

Comment:
this doesn't need an else since the previous if raises or returns. I think taking it out might help make the code flow a bit easier to read.